src/mock/ray/core_worker/memory_store.h

```diff
@@ -0,0 +1,51 @@
+// Copyright 2024 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#pragma once
+
+#include "ray/core_worker/store_provider/memory_store/memory_store.h"
+
+namespace ray::core {
+
+// Creates a CoreWorkerMemoryStore with a thread and a event loop hosting the callbacks.
+// Should only be used in CPP tests.
+class DefaultCoreWorkerMemoryStoreWithThread : public CoreWorkerMemoryStore {
+ public:
+  static std::unique_ptr<DefaultCoreWorkerMemoryStoreWithThread> Create() {
+    std::unique_ptr<InstrumentedIOContextWithThread> io_context =
+        std::make_unique<InstrumentedIOContextWithThread>(
+            "DefaultCoreWorkerMemoryStoreWithThread");
+    return std::unique_ptr<DefaultCoreWorkerMemoryStoreWithThread>(
+        new DefaultCoreWorkerMemoryStoreWithThread(std::move(io_context)));
+  }
+
+  static std::shared_ptr<DefaultCoreWorkerMemoryStoreWithThread> CreateShared() {
+    std::unique_ptr<InstrumentedIOContextWithThread> io_context =
+        std::make_unique<InstrumentedIOContextWithThread>(
+            "DefaultCoreWorkerMemoryStoreWithThread");
+    return std::shared_ptr<DefaultCoreWorkerMemoryStoreWithThread>(
```

Comment:
Why not `std::make_shared`? Your implementation leads to two memory allocations instead of one.

Reply:
because the ctor is private and thus not visible to make_shared. you can make a private tag type trick but for sake of simplicity I did not do that.